File: .github/set-secrets.py
#!/usr/bin/env python3
"""
GitHub Secrets Setup Script
Automatically sets GitHub Secrets from .env file
"""

import os
import sys
import subprocess
import json
import base64
from pathlib import Path
from nacl import encoding, public

def get_env_vars():
    """Read environment variables from .env file"""
    env_file = Path("codescan-pro/.env")
    if not env_file.exists():
        print("❌ codescan-pro/.env file not found!")
        sys.exit(1)
    
    env_vars = {}
    with open(env_file, 'r') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#') and '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                if value:  # Only add non-empty values
                    env_vars[key] = value
    
    return env_vars

def encrypt_secret(public_key: str, secret_value: str) -> str:
    """Encrypt a secret using GitHub's public key"""
    public_key_bytes = base64.b64decode(public_key)
    public_key_obj = public.PublicKey(public_key_bytes)
    sealed_box = public.SealedBox(public_key_obj)
    encrypted = sealed_box.encrypt(secret_value.encode('utf-8'))
    return base64.b64encode(encrypted).decode('utf-8')

def get_github_public_key(repo_owner: str, repo_name: str, token: str):
    """Get GitHub repository's public key for encryption"""
    import urllib.request
    import urllib.error
    
    url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/actions/secrets/public-key"
    req = urllib.request.Request(url)
    req.add_header("Authorization", f"token {token}")
    req.add_header("Accept", "application/vnd.github.v3+json")
    
    try:
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read())
            return data['key'], data['key_id']
    except urllib.error.HTTPError as e:
        print(f"❌ Error getting public key: {e}")
        if e.code == 401:
            print("   Authentication failed. Check your GitHub token.")
        elif e.code == 404:
            print("   Repository not found or you don't have access.")
        sys.exit(1)

def set_github_secret(repo_owner: str, repo_name: str, secret_name: str, secret_value: str, token: str):
    """Set a GitHub secret using the API"""
    import urllib.request
    import urllib.error
    
    # Get public key
    public_key, key_id = get_github_public_key(repo_owner, repo_name, token)
    
    # Encrypt secret
    encrypted_value = encrypt_secret(public_key, secret_value)
    
    # Set secret
    url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/actions/secrets/{secret_name}"
    data = json.dumps({
        "encrypted_value": encrypted_value,
        "key_id": key_id
    }).encode('utf-8')
    
    req = urllib.request.Request(url, data=data, method='PUT')
    req.add_header("Authorization", f"token {token}")
    req.add_header("Accept", "application/vnd.github.v3+json")
    req.add_header("Content-Type", "application/json")
    
    try:
        with urllib.request.urlopen(req) as response:
            return True
    except urllib.error.HTTPError as e:
        print(f"❌ Error setting {secret_name}: {e}")
        if e.code == 401:
            print("   Authentication failed. Check your GitHub token.")
        return False

def main():
    repo_owner = "Sharjeel-Saleem-06"
    repo_name = "CodeSentinel"
    
    print("🔐 GitHub Secrets Setup")
    print("=" * 50)
    
    # Try to get token from environment or GitHub CLI
    token = os.environ.get('GITHUB_TOKEN')
    
    if not token:
        # Try GitHub CLI
        try:
            result = subprocess.run(['gh', 'auth', 'token'], capture_output=True, text=True, check=True)
            token = result.stdout.strip()
            print("✅ Using GitHub CLI token")
        except (subprocess.CalledProcessError, FileNotFoundError):
            print("❌ GitHub token not found!")
            print("\nPlease provide a GitHub Personal Access Token:")
            print("1. Go to: https://github.com/settings/tokens")
            print("2. Generate new token (classic) with 'repo' scope")
            print("3. Set it as environment variable: export GITHUB_TOKEN=your_token")
            print("   Or run: GITHUB_TOKEN=your_token python3 .github/set-secrets.py")
            sys.exit(1)
    
    # Get Netlify secrets (user needs to provide these)
    print("\n📋 Netlify Configuration Required:")
    netlify_token = os.environ.get('NETLIFY_AUTH_TOKEN')
    netlify_site_id = os.environ.get('NETLIFY_SITE_ID')
    
    if not netlify_token:
        print("⚠️  NETLIFY_AUTH_TOKEN not set. Skipping Netlify secrets.")
        print("   Get it from: https://app.netlify.com/user/applications")
    if not netlify_site_id:
        print("⚠️  NETLIFY_SITE_ID not set. Skipping Netlify secrets.")
        print("   Get it from: Netlify Dashboard → Site settings → General")
    
    # Read .env file
    print("\n📖 Reading codescan-pro/.env...")
    env_vars = get_env_vars()
    print(f"✅ Found {len(env_vars)} environment variables")
    
    # Set secrets
    print("\n🔑 Setting GitHub Secrets...")
    success_count = 0
    fail_count = 0
    
    # Set API key secrets
    for key, value in env_vars.items():
        if key.startswith('VITE_'):
            print(f"   Setting {key}...", end=' ')
            if set_github_secret(repo_owner, repo_name, key, value, token):
                print("✅")
                success_count += 1
            else:
                print("❌")
                fail_count += 1
    
    # Set Netlify secrets if provided
    if netlify_token:
        print(f"   Setting NETLIFY_AUTH_TOKEN...", end=' ')
        if set_github_secret(repo_owner, repo_name, "NETLIFY_AUTH_TOKEN", netlify_token, token):
            print("✅")
            success_count += 1
        else:
            print("❌")
            fail_count += 1
    
    if netlify_site_id:
        print(f"   Setting NETLIFY_SITE_ID...", end=' ')
        if set_github_secret(repo_owner, repo_name, "NETLIFY_SITE_ID", netlify_site_id, token):
            print("✅")
            success_count += 1
        else:
            print("❌")
            fail_count += 1
    
    print("\n" + "=" * 50)
    print(f"✅ Successfully set: {success_count} secrets")
    if fail_count > 0:
        print(f"❌ Failed: {fail_count} secrets")
    print("\n🎉 Done! Your GitHub Actions workflow will now use these secrets.")

if __name__ == "__main__":
    # Check if nacl is installed
    try:
        from nacl import encoding, public
    except ImportError:
        print("❌ PyNaCl library required!")
        print("   Install it: pip3 install pynacl")
        sys.exit(1)
    
    main()

File: .github/set-secrets.js
#!/usr/bin/env node

/**
 * GitHub Secrets Setup Script
 * Sets all secrets from .env file to GitHub repository
 * 
 * Usage:
 *   GITHUB_TOKEN=your_token node .github/set-secrets.js
 * 
 * Or with GitHub CLI authenticated:
 *   node .github/set-secrets.js
 */

const fs = require('fs');
const { execSync } = require('child_process');
const https = require('https');

const REPO_OWNER = 'Sharjeel-Saleem-06';
const REPO_NAME = 'CodeSentinel';
const ENV_FILE = 'codescan-pro/.env';

// Get GitHub token
function getToken() {
    // Try environment variable first
    if (process.env.GITHUB_TOKEN) {
        return process.env.GITHUB_TOKEN;
    }
    
    // Try GitHub CLI
    try {
        return execSync('gh auth token', { encoding: 'utf-8' }).trim();
    } catch (e) {
        return null;
    }
}

// Read .env file
function readEnvFile() {
    if (!fs.existsSync(ENV_FILE)) {
        console.error(`❌ ${ENV_FILE} not found!`);
        process.exit(1);
    }
    
    const content = fs.readFileSync(ENV_FILE, 'utf-8');
    const vars = {};
    
    content.split('\n').forEach(line => {
        line = line.trim();
        if (line && !line.startsWith('#') && line.includes('=')) {
            const [key, ...valueParts] = line.split('=');
            const value = valueParts.join('=').trim().replace(/^["']|["']$/g, '');
            if (value && key.startsWith('VITE_')) {
                vars[key.trim()] = value;
            }
        }
    });
    
    return vars;
}

// Make HTTPS request
function makeRequest(options, data) {
    return new Promise((resolve, reject) => {
        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                resolve({ statusCode: res.statusCode, body });
            });
        });
        
        req.on('error', reject);
        
        if (data) {
            req.write(JSON.stringify(data));
        }
        
        req.end();
    });
}

// Get public key
async function getPublicKey(token) {
    const options = {
        hostname: 'api.github.com',
        path: `/repos/${REPO_OWNER}/${REPO_NAME}/actions/secrets/public-key`,
        method: 'GET',
        headers: {
            'Authorization': `token ${token}`,
            'Accept': 'application/vnd.github.v3+json',
            'User-Agent': 'Node.js'
        }
    };
    
    const response = await makeRequest(options);
    
    if (response.statusCode !== 200) {
        throw new Error(`Failed to get public key: ${response.statusCode}`);
    }
    
    return JSON.parse(response.body);
}

// Encrypt secret (requires libsodium - we'll use a simple approach)
// Note: For production, use proper encryption library
async function encryptSecret(publicKey, secretValue, keyId) {
    // This is a simplified version - in production use proper encryption
    // For now, we'll use GitHub CLI if available, or prompt user
    try {
        // Try using gh secret set which handles encryption
        return { encrypted: true, method: 'gh-cli' };
    } catch (e) {
        throw new Error('Encryption requires GitHub CLI or PyNaCl library');
    }
}

// Set secret using GitHub CLI (simplest method)
function setSecretWithCLI(secretName, secretValue) {
    try {
        execSync(`echo -n "${secretValue}" | gh secret set ${secretName} --repo ${REPO_OWNER}/${REPO_NAME}`, {
            stdio: 'inherit'
        });
        return true;
    } catch (e) {
        return false;
    }
}

// Main function
async function main() {
    console.log('🔐 GitHub Secrets Setup');
    console.log('='.repeat(50));
    console.log('');
    
    // Get token
    const token = getToken();
    if (!token) {
        console.error('❌ GitHub token not found!');
        console.error('');
        console.error('Option 1: Set GITHUB_TOKEN environment variable');
        console.error('   GITHUB_TOKEN=your_token node .github/set-secrets.js');
        console.error('');
        console.error('Option 2: Authenticate GitHub CLI');
        console.error('   gh auth login');
        console.error('   node .github/set-secrets.js');
        process.exit(1);
    }
    
    console.log('✅ GitHub token found');
    
    // Read .env
    console.log(`📖 Reading ${ENV_FILE}...`);
    const envVars = readEnvFile();
    console.log(`✅ Found ${Object.keys(envVars).length} environment variables`);
    console.log('');
    
    // Check if gh CLI is available
    let useCLI = false;
    try {
        execSync('gh --version', { stdio: 'ignore' });
        useCLI = true;
        console.log('✅ Using GitHub CLI for encryption');
    } catch (e) {
        console.log('⚠️  GitHub CLI not found - will use API (requires PyNaCl)');
    }
    console.log('');
    
    // Set secrets
    console.log('🔑 Setting secrets...');
    let success = 0;
    let failed = 0;
    
    for (const [key, value] of Object.entries(envVars)) {
        process.stdout.write(`   Setting ${key}... `);
        
        if (useCLI) {
            if (setSecretWithCLI(key, value)) {
                console.log('✅');
                success++;
            } else {
                console.log('❌');
                failed++;
            }
        } else {
            // Would need to implement API method with encryption
            console.log('⚠️  (Requires GitHub CLI)');
            failed++;
        }
    }
    
    console.log('');
    console.log('='.repeat(50));
    console.log(`✅ Successfully set: ${success} secrets`);
    if (failed > 0) {
        console.log(`❌ Failed: ${failed} secrets`);
    }
    console.log('');
    console.log('📋 Don\'t forget to set Netlify secrets:');
    console.log('   NETLIFY_AUTH_TOKEN');
    console.log('   NETLIFY_SITE_ID');
    console.log('');
    console.log('🎉 Done!');
}

main().catch(console.error);
